### DISCUSSION TITLE
Recommended way to mock Kingfisher in tests?

### DISCUSSION THREAD

**Discussion by CraigSiemens** [onevcat/Kingfisher#1939] [Q&A] [answered]
I'm trying to write some tests for a view in the app that contains an image that's loaded by Kingfisher. I have a couple options for the type of tests I could write, but I'm not finding a obvious way to mock out Kingfisher for use in tests.

If I write a unit test, I can create a subclass of `ImageDownloader` and override `downloadImage` to check if it's passed the correct values. But I can't then manually call the completion with a mock image since `ImageLoadingResult` doesn't have a public init.

If I write a snapshot test, I can pass it a file url but that is asynchronously loaded after the snapshot is taken. Subclassing `ImageDownloader` to synchronously return an image also isn't an option since the result also doesn't have a public init.

**Comment by onevcat** [comment]
I am not quite sure what kind of test you are trying to add, or say, can you paste some pseudo-code so we can have a more concrete context to discuss.

> If I write a snapshot test, I can pass it a file url but that is asynchronously loaded after the snapshot is taken

If a file URL is an option for your testing, then maybe [this `load​Disk​File​Synchronously`](https://kingfisher.onevcat.com/kingfisherparsedoptionsinfo/#kingfisherparsedoptionsinfo.loaddiskfilesynchronously) can help you out:

```swift
KF.url(fileURL)
  .loadDiskFileSynchronously()
  .set(to: imageView)

// Or if you prefer the image view extension:
imageView.kf.setImage(with: fileURL, options: [.loadDiskFileSynchronously])
```

**Comment by CraigSiemens** [comment, marked as answer]
We're using https://github.com/pointfreeco/swift-snapshot-testing to take snapshots of our views in some of our tests. The goal was to have two tests, one to make sure the view had the correct appearance when the image hadn't loaded yet, and another for after the image did load.

What I ended up doing is making a cache in the test and setting the default options on kingfisher to to only use that cache. Something like this.

```swift 
describe("appearance") {
    beforeEach {
        imageCache = ImageCache(name: self.name)
	
        KingfisherManager.shared.defaultOptions = [
            .onlyFromCache,
            .targetCache(imageCache)
        ]
    }

    afterEach {
        KingfisherManager.shared.defaultOptions = []
    }

    context("when the image hasn't loaded yet") {
        beforeEach {
            imageCache.clearMemoryCache()
            view.configure(with: viewModel)
        }
        
        it("should have a dark background") {
            expect(view).to(haveMatchingSnapshot())
        }
    }

    context("when the image has loaded") {
        beforeEach {
            imageCache.store(
                .placeholder,
                forKey: expectedURL.cacheKey,
                toDisk: false
            )
	    
            view.configure(with: viewModel)
        }
        
        it("should show the image") {
            expect(view).to(haveMatchingSnapshot())
        }
    }
}
```

**Comment by tylerlantern** [comment]
Update 
```
    let cache = ImageCache(name: "VoucherDetailComponentTests")
    KingfisherManager.shared.defaultOptions = [
      .onlyFromCache,
      .targetCache(cache)
    ]
    let options : [KingfisherOptionsInfoItem] = [.targetCache(cache)]
     cache.store(
       KFCrossPlatformImage(systemName: "fireplace.fill")!,
       forKey:  Notify.Voucher.Place.mock.logo!.cacheKey,
       options: KingfisherParsedOptionsInfo(KingfisherOptionsInfo(options))
     )
```
where Notify.Voucher.Place.mock.logo is URL Type
I used snapshot testing from [ttps://github.com/pointfreeco/swift-snapshot-testing](https://github.com/pointfreeco/swift-snapshot-testing) to test image

**Comment by MartinStrambach** [comment]
FYI this approach was broken by https://github.com/onevcat/Kingfisher/pull/2036 in 7.6.2 if you use `startLoadingBeforeViewAppear`
